### PR TITLE
Additions for evolutions 4

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Work around [this](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=64095) bug in GCC
- The `Tag::ElementMap` now actually holds an `ElementMap` instead of a `unique_ptr<CoordinateMapBase>`
- Add tag for computing the coordinates from a coordinate map tag and source coordinates
- Add an `DomainCreator` options tag. @wthrowe do we have a method of testing these tags that I've forgotten about?
- Fix a bug in Variables
- Changes to FluxCommunication so it interfaces with the fluxes from ScalarWave. It's likely this will need to be adapted further when we have dynamical backgrounds.

These are the last changes to FluxCommunication.hpp that I need to get evolutions to work.

**Edit:** there is one more thing I need to change in FluxCommunication.hpp, which I'm changing now. It's that dt_vars is not added by them, but rather is mutated in the DataBox.

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

### CODE REVIEW

**Comment by nilsdeppe** [comment]
I've added two commits that make the remaining changes for FluxCommunication.hpp to be identical to what I have on the evolution branch.

File: src/Parallel/Main.hpp [hunk 1]
```diff
@@ -117,7 +117,7 @@ Main<Metavariables>::Main(CkArgMsg* msg) noexcept
                "Input file name");
         },
         [&command_line_options](std::true_type /*meta*/, auto /*mv*/,
-                                auto... /*unused*/) {
+                                ... /*unused*/) {
```
Comment (kidder) [inline src/Parallel/Main.hpp]: general query: does any test test the four possibilities of the enclosing make_overloader?

File: tests/Unit/Domain/Test_CoordinatesTag.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "DataStructures/DataBox.hpp"
```
Comment (kidder) [inline tests/Unit/Domain/Test_CoordinatesTag.cpp]: add includes for size_t, Index, Frames,

File: src/DataStructures/Variables.hpp [hunk 3]
```diff
@@ -402,8 +402,9 @@ Variables<tmpl::list<Tags...>>::Variables(
     : variable_data_impl_(std::move(rhs.variable_data_impl_)),
       variable_data_(std::move(rhs.variable_data_)),
       size_(rhs.size()),
-      number_of_grid_points_(rhs.number_of_grid_points()),
-      reference_variable_data_(std::move(rhs.reference_variable_data_)) {}
+      number_of_grid_points_(rhs.number_of_grid_points()) {
+  add_reference_variable_data(typelist<Tags...>{});
```
Comment (kidder) [inline src/DataStructures/Variables.hpp]: why didn't an existing test catch this bug?  can the test be updated?

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp [hunk 4]
```diff
@@ -72,13 +72,30 @@ struct ComputeBoundaryFlux {
...
+      const tnsr::i<DataVector, Dim, Frame>& unit_face_normal,
+      tmpl::list<ArgumentTags...> /*meta*/) noexcept {
+    System::normal_dot_fluxes::apply(
+        make_not_null(&get<NormalDotFluxTags>(*boundary_flux))...,
+        get<ArgumentTags>(variables)..., unit_face_normal);
+  }
+
+  template <typename Flux, typename... NumericalFluxTags, typename... Tags,
+            typename... ArgumentTags, typename... Args>
+  static auto apply_normal_dot_numerical_flux(
```
Comment (kidder) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp]: auto = void?

File: tests/Unit/Domain/Test_CoordinatesTag.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  const auto box = db::create<
+      db::AddTags<Tags::Extents<Dim>, map_tag>,
+      db::AddComputeItemsTags<
+          Tags::LogicalCoordinates<Dim>,
+          Tags::Coordinates<map_tag, Tags::LogicalCoordinates<Dim>>>>(
+      extents, ElementMap<Dim, Frame::Grid>(
+                   ElementId<Dim>(0),
+                   make_coordinate_map_base<Frame::Logical, Frame::Grid>(map)));
+  CHECK_ITERABLE_APPROX(
+      (db::get<Tags::Coordinates<map_tag, Tags::LogicalCoordinates<Dim>>>(box)),
```
Comment (kidder) [inline tests/Unit/Domain/Test_CoordinatesTag.cpp:29]: will I have to fetch the coordinates this way?  I would like to get them with something like db::get<Tags::Coordinates<Frame>>(box)

**Review by kidder — CHANGES_REQUESTED** [review]
more to come, I didn't make it all the way through

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp (hunk 4 above)
Comment (nilsdeppe) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp]: Yep

File: src/Parallel/Main.hpp (hunk 1 above)
Comment (nilsdeppe) [reply src/Parallel/Main.hpp]: I don't know, I'd have to look. @wthrowe might know

File: tests/Unit/Domain/Test_CoordinatesTag.cpp (hunk 5 above)
Comment (nilsdeppe) [reply tests/Unit/Domain/Test_CoordinatesTag.cpp:29]: Yes, currently the long way is necessary. We could alternatively have things like `Tags::InertialCoordinates` etc. Another possibility might be to just template on the target frame. This would require making a canonical decision on source frame, and map. That could be a bit tricky with ElementMap and dual frames. I'd have to think about it.

File: src/Parallel/Main.hpp (hunk 1 above)
Comment (wthrowe) [reply src/Parallel/Main.hpp]: Nils rewrote these tests so I'm not sure, but I think only the first case is tested.  The third case is a static_assert so that one's fine.  The fourth would be straightforward to test by creating a test with no options.  The second (the one changed here) is hard because it requires parsing an input file.  I guess it could be tested with a failure test.

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp [hunk 6]
```diff
@@ -284,10 +287,12 @@ SPECTRE_TEST_CASE(
 
   const Element<2> element(self_id, {});
 
-  auto map = make_coordinate_map_base<Frame::Logical, Frame::Grid>(
-      CoordinateMaps::ProductOf2Maps<CoordinateMaps::AffineMap,
-                                     CoordinateMaps::AffineMap>(
-          {-1., 1., 3., 7.}, {-1., 1., -2., 4.}));
+  auto map = ElementMap<2, Frame::Grid>(
+      ElementId<2>{0},
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp]: `self_id`.  The other one *should* also be `self_id`, but if you make that change I think the test may break.  You can check if it breaks anything if you like, but if it does just leave it and I'll fix it in my cleanup of that test.

File: src/Domain/Tags.hpp [hunk 7]
```diff
@@ -31,6 +32,17 @@ struct Logical;
...
 }  // namespace Frame
 
+namespace OptionTags {
+/// \ingroup DataBoxTagsGroup
+/// \ingroup ComputationalDomainGroup
+/// The input file tag for the DomainCreator to use
+template <size_t Dim, typename TargetFrame>
+struct DomainCreator {
+  using type = std::unique_ptr<::DomainCreator<Dim, TargetFrame>>;
+  static constexpr const char* const help = {"The domain to create initially"};
```
Comment (wthrowe) [inline src/Domain/Tags.hpp]: `OptionString`

File: src/NumericalAlgorithms/DiscontinuousGalerkin/LiftFlux.hpp [hunk 8]
```diff
@@ -31,11 +31,12 @@ namespace dg {
 ///
 /// \note The result is still provided only on the boundary grid.  The
 /// values away from the boundary are zero and are not stored.
-template <typename Tags>
-Variables<Tags> lift_flux(const Variables<Tags>& local_flux,
-                          Variables<Tags> numerical_flux,
-                          const size_t extent_perpendicular_to_boundary,
-                          DataVector magnitude_of_face_normal) noexcept {
+template <typename NormalDotFluxTags, typename NormalDotNumericalFluxTags>
+Variables<NormalDotNumericalFluxTags> lift_flux(
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/LiftFlux.hpp]: Shouldn't the return type be the time derivative of the evolved variables or something like that?

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp [hunk 9]
```diff
@@ -145,57 +161,51 @@ struct ComputeBoundaryFlux {
 
       // Using this instead of auto prevents incomprehensible errors
       // if the return type of compute_flux is wrong.
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp]: Update this comment and move it to the right place (if it is still relevant).

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp [hunk 10]
```diff
@@ -145,57 +161,51 @@ struct ComputeBoundaryFlux {
...
+
+      db::item_type<
+          db::add_tag_prefix<Tags::NormalDotNumericalFlux, variables_tag>>
+          normal_dot_numerical_fluxes(face_normal.begin()->size(), 0.0);
+      apply_normal_dot_numerical_flux(
+          make_not_null(&normal_dot_numerical_fluxes),
+          normal_dot_numerical_flux_computer, self_data, neighbor_data,
+          typename std::decay_t<decltype(
+              normal_dot_numerical_flux_computer)>::argument_tags{},
+          unit_face_normal);
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp]: Can this depend on unit_face_normal?  I thought we decided it couldn't when we were discussing fluxes in GH.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp [hunk 11]
```diff
@@ -110,8 +110,9 @@ struct ComputeBoundaryFlux {
     static_assert(cpp17::is_same_v<System, typename Metavariables::system>,
                   "Inconsistent systems");
     constexpr const size_t volume_dim = System::volume_dim;
-    using variables_tag = typename System::variables_tag;
-    using dt_variables_tag = typename System::dt_variables_tag;
+    using variables_tag = Tags::Variables<typename System::variables_tags>;
+    using dt_variables_tag = Tags::dt<Tags::Variables<
+        db::wrap_tags_in<Tags::dt, typename System::variables_tags>>>;
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp]: `using dt_variables_tag = db::add_tag_prefix<Tags::dt, variables_tag>;`

**Review by wthrowe — CHANGES_REQUESTED** [review]
Haven't looked at the last commit.

File: src/Parallel/Main.hpp (hunk 1 above)
Comment (kidder) [reply src/Parallel/Main.hpp]: we can easily test this once we have an example input file parsing test

Comment (kidder) [inline src/Parallel/Main.hpp]: add a comment about why the change is made

File: src/NumericalAlgorithms/DiscontinuousGalerkin/LiftFlux.hpp (hunk 8 above)
Comment (kidder) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/LiftFlux.hpp]: `return Variabes<dt...>(std::move(num_flux))`

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp [hunk 12]
```diff
@@ -193,18 +193,12 @@ struct ComputeBoundaryFlux {
...
-      const auto lifted_data = dg::lift_flux(
-          tuples::get<normal_flux_tag>(self_data),
-          std::move(numerical_flux),
-          extents[dimension],
-          std::move(magnitude_of_face_normal));
-
-      // Convert tags.  This will be replaced by expanded Variables
-      // functionality.  See #319.
-      std::decay_t<decltype(dt_vars)> retagged_lifted_data = 1. * lifted_data;
+      std::decay_t<decltype(dt_vars)> lifted_data(dg::lift_flux(
```
Comment (kidder) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp]: auto after change to a previous commit

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp (hunk 10 above)
Comment (kidder) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp]: just send the normal

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp [hunk 13]
```diff
@@ -72,13 +72,30 @@ struct ComputeBoundaryFlux {
...
   using inbox_tags = tmpl::list<FluxesTag>;
 
-  template <typename Flux, typename... Tags, typename... ArgumentTags>
-  static auto apply_flux(const Flux& flux,
-                         const tuples::TaggedTuple<Tags...>& self_data,
-                         const tuples::TaggedTuple<Tags...>& neighbor_data,
-                         tmpl::list<ArgumentTags...> /*meta*/) noexcept {
-    return flux(tuples::get<ArgumentTags>(self_data)...,
-                tuples::get<ArgumentTags>(neighbor_data)...);
+  template <typename... NormalDotFluxTags, typename... VariablesTags,
```
Comment (kidder) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp]: make a new action for non-conservative systems and keep the current action for conservative systems

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp [hunk 14]
```diff
@@ -33,10 +33,10 @@ struct Var : db::DataBoxTag {
 
 struct System {
   static constexpr const size_t volume_dim = 2;
-  using variables_tag = Tags::Variables<tmpl::list<Var>>;
-  using dt_variables_tag = Tags::dt<Tags::Variables<tmpl::list<Tags::dt<Var>>>>;
+  using variables_tags = tmpl::list<Var>;
```
Comment (kidder) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp]: write things in terms of variables_tag instead of variables_tags

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp [hunk 15]
```diff
@@ -122,6 +123,9 @@ SPECTRE_TEST_CASE("Unit.DiscontinuousGalerkin.Actions.FluxCommunication",
   const CoordinateMaps::AffineMap xi_map{-1., 1., 3., 7.};
   const CoordinateMaps::AffineMap eta_map{-1., 1., -2., 4.};
 
+  using dt_variables_tag = Tags::dt<Tags::Variables<
```
Comment (kidder) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp]: add_tag_prefix

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp [hunk 16]
```diff
@@ -279,14 +286,16 @@ SPECTRE_TEST_CASE("Unit.DiscontinuousGalerkin.Actions.FluxCommunication",
...
-      get<Tags::dt<Var>>(db::get<System::dt_variables_tag>(received_box)).get(),
+      get<Tags::dt<Var>>(db::get<dt_variables_tag>(received_box)).get(),
       xi_lift * xi_boundaries + eta_lift * eta_boundaries);
 }
 
 SPECTRE_TEST_CASE(
     "Unit.DiscontinuousGalerkin.Actions.FluxCommunication.NoNeighbors",
     "[Unit][NumericalAlgorithms][Actions]") {
   ActionTesting::ActionRunner<Metavariables> runner{{}};
+  using dt_variables_tag = Tags::dt<Tags::Variables<
```
Comment (kidder) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp]: add_tags_prefix

File: src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp [hunk 17]
```diff
@@ -121,8 +121,6 @@ struct ComputeBoundaryFlux {
         db::get<Tags::HistoryBoundaryVariables<
                   Direction<volume_dim>, variables_tag>>(box);
 
-    auto dt_vars = db::item_type<dt_variables_tag>(extents.product(), 0.);
```
Comment (kidder) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:106]: add the part where this is zeroed to the PR

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
I've updated this addressing the comments. The PR depends on #468. I've removed the changes to the FluxCommunication stuff and will add the new Actions in a separate PR to keep PRs smaller.

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@wthrowe can you take a final look at this PR

File: src/Domain/Tags.hpp [hunk 18]
```diff
@@ -31,6 +33,17 @@ struct Logical;
 struct Inertial;
 }  // namespace Frame
 
+namespace OptionTags {
+/// \ingroup DataBoxTagsGroup
```
Comment (wthrowe) [inline src/Domain/Tags.hpp]: Not a DataBoxTag.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/LiftFlux.hpp [hunk 19]
```diff
@@ -48,6 +52,8 @@ Variables<Tags> lift_flux(const Variables<Tags>& local_flux,
   auto lifted_data = std::move(numerical_flux);
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/LiftFlux.hpp:47]: The comment for this line is no longer correct.

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
I've pushed (rebased) updates to the commits for DomainCreator tag and lift_flux return type. I also added 2 new small commits:

1. Add the OptionTagsGroup Doxygen group
2. Add the relevant tags in Time/Tags.hpp to that group. They so far were in the non-existent OptionTags group.

The rendered documentation is available [here](http://nilsdeppe.com/spectre/pr_465/)

**Review by wthrowe — DISMISSED** [review]

**Review by kidder — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
This needs to be rebased after #474 is merged.

**Comment by nilsdeppe** [comment]
I rebased this on top of #474 so that once that is merged this can be merged too

Edit:

I did this to try deal with the long macOS build times.

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
because of long queues for Travis mac builds, I tested this on my laptop
